File: NumberOfBSTFromArray/NumberOfBSTFromArray.java
/*
https://www.geeksforgeeks.org/problems/number-of-bst-from-array/1
*/

class Solution {
    
    public int nCr(int N, int R){
        if(R>N-R) R=N-R;
        int res=1;
        for(int i=0; i<R; ++i){
            res*=(N-i);
            res/=(i+1);
        }
        return res;
    }
    
    public ArrayList<Integer> computeCatalan(int N){
        ArrayList<Integer>catalan=new ArrayList<>();
        for(int i=0; i<=N; ++i){
            catalan.add(0);
        }
        for(int i=0; i<=N; ++i){
            int c=nCr(2*i, i);
            catalan.set(i, c/(i+1));
        }
        return catalan;
    }
    
    public ArrayList<Integer> countBSTs(int[] arr) {
        int lengthOfArr=arr.length;
        ArrayList<Integer>sortedArr=new ArrayList<>();
        for(int i=0; i<lengthOfArr; ++i){
            sortedArr.add(arr[i]);
        }
        Collections.sort(sortedArr);
        HashMap<Integer, Integer>rankMap=new HashMap<>();
        for(int i=0; i<lengthOfArr; ++i){
            rankMap.put(sortedArr.get(i), i);
        }
        ArrayList<Integer>catalan=computeCatalan(lengthOfArr);
        ArrayList<Integer>resultantList=new ArrayList<>();
        for(int vals: arr){
            int L=rankMap.get(vals), R=lengthOfArr-1-L, total=catalan.get(L)*catalan.get(R);
            resultantList.add(total);
        }
        return resultantList;
    }
}

File: NumberOfBSTFromArray/NumberOfBSTFromArray.cpp
/*
https://www.geeksforgeeks.org/problems/number-of-bst-from-array/1
*/

class Solution {
  public:
  
    int nCr(int n, int r){
        if (r > n - r) r = n - r;
        int res = 1;
        for (int i = 0; i < r; ++i) {
            res *= (n - i);
            res /= (i + 1);
        }
        return res;
    }
  
    vector<int> computeCatalan(int n){
        vector<int> catalan(n + 1);
        for (int i = 0; i <= n; ++i) {
            int c = nCr(2 * i, i);
            catalan[i] = c / (i + 1);
        }
        return catalan;
    }
  
    vector<int> countBSTs(vector<int>& arr) {
        int n = arr.size();
        vector<int> sortedArr = arr;
        sort(sortedArr.begin(), sortedArr.end());
        unordered_map<int, int> rankMap;
        for (int i = 0; i < n; ++i) {
            rankMap[sortedArr[i]] = i;
        }
        vector<int> catalan = computeCatalan(n);
        vector<int> result;
        result.reserve(n);
        for (int val : arr) {
            int L = rankMap[val];
            int R = n - 1 - L;
            int total = catalan[L] * catalan[R];
            result.push_back(total);
        }

        return result;
    }
};